### DISCUSSION TITLE
Use webhooks

### DISCUSSION THREAD

**Discussion by Cha0s2nd** [rainepretorius/olarm-ha-integration#80] [Ideas]
The webhooks should resolve the rate limit issues and might give more real-time updates.

Can't be that hard, unfortunately I'm not a python expert.

**Comment by rainepretorius** [comment]
It is not possible for an integration. The whole package would have to be redeveloped and packaged as a Home Assistant add-on.

**Reply by Cha0s2nd** [reply]
So at risk of sounding stupid and being pushy, can you explain why you say that?
According to the doumentation pushing apis are a thing, and there's a webhook section for the homeassistant api.
I'm obviously missing some context but where do these not fit together?

https://developers.home-assistant.io/docs/api/native-app-integration/sending-data/
https://developers.home-assistant.io/docs/integration_fetching_data
![image](https://github.com/rainepretorius/olarm-ha-integration/assets/8254451/782e78fc-ee4e-4b77-9be5-26dc576d44de)

**Reply by Cha0s2nd** [reply]
I think I see why the webhook part of hass is pretty locked down, it's a pity.

**Comment by rainepretorius** [comment]
Let me try to clarify the distinction between the two:

Pushing Integration:

In a pushing integration, external devices or services send updates or data to Home Assistant in real-time or near real-time.
This means that when the state of a sensor changes, the sensor itself actively informs Home Assistant about the change.
Pushing integrations typically result in lower latency because the data is immediately available to Home Assistant for processing and automation triggers.
Polling Integration:

In a polling integration, Home Assistant actively queries or requests data from external devices or services at regular intervals.
Home Assistant periodically checks the state of the devices or services to see if there are any updates.
Polling integrations can introduce more latency because there may be a delay between the actual state change of a sensor and when Home Assistant polls for updates.
The choice between pushing and polling integrations depends on various factors, including the capabilities of your devices, the availability of APIs, and your specific automation needs. Pushing integrations are generally preferred when real-time or low-latency updates are crucial, but not all devices or services support pushing data.

**Reply by Cha0s2nd** [reply]
So this I get, and the webhook allows a push mechanism, but I don't get how the push integrations work with no exposed endpoints.
Can't the integration just use the internal webserver to give a url, similar to this:
https://developers.home-assistant.io/docs/api/rest/

![image](https://github.com/rainepretorius/olarm-ha-integration/assets/8254451/1f09bc72-6738-40d9-9004-c100f0f7cb6f)

The only incoming thing that you can control the data received seems to be the Automation Webhook triggers

**Reply by rainepretorius** [reply]
I can't entirely agree with your assertion because Home Assistant's API requires authentication. There is no native push integration available. The discussion primarily pertains to how sensor data is configured within the integration.

As for your inquiry regarding the exposure of endpoints through Home Assistant's internal webserver, it's important to note that the internal webserver primarily serves the Home Assistant user interface and provides API endpoints for internal communication. This webserver is not typically utilized for receiving data from external sources in the same manner as webhooks. This separation is vital to maintain security, as exposing internal API endpoints directly to the internet can introduce significant security risks. Therefore, webhooks are generally the preferred method for securely handling external data.

**Reply by Cha0s2nd** [reply]
I think you are starting to see where I'm going and it's more around what they are allowing integrations to do. (exposing anything is users own risk if it's webhook or api).
It's all there, just not how it will make sense for what this discussion is trying to achieve. It does make more sense why it will be an add-on now, having their own docker containers does isolate it way better.

To be honest you can use the current integration as is, with a very high interval and then set up a webhook triggered automation.
It should give you real-time events, but it's not worth the effort I would say.
